archive/type.go

```diff
@@ -0,0 +1,216 @@
+			}
+		case *zng.TypeUnion:
+			if !container {
+				return ErrSyntax
+			}
+			if err := t.union(v, body); err != nil {
+				return err
+			}
+		default:
+			if container {
+				return ErrSyntax
+			}
+			t.value(body)
+		}
+	}
+	return nil
+}
+
+func (t *TypeIndexer) union(typ *zng.TypeUnion, body zcode.Bytes) error {
+	if len(body) == 0 {
+		return nil
+	}
+	it := zcode.Iter(body)
+	v, container, err := it.Next()
+	if err != nil {
+		return err
+	}
+	if container {
+		return ErrSyntax
+	}
+	index := zcode.DecodeCountedUvarint(v)
+	inner, err := typ.TypeIndex(int(index))
+	if err != nil {
+		return err
+	}
+	body, container, err = it.Next()
+	if err != nil {
+		return err
+	}
+	switch v := zng.AliasedType(inner).(type) {
+	case *zng.TypeRecord:
+		if !container {
+			return ErrSyntax
+		}
+		if err := t.record(v, body); err != nil {
+			return err
+		}
+	case *zng.TypeArray:
+		if !container {
+			return ErrSyntax
+		}
+		if err := t.vector(v, body); err != nil {
+			return err
+		}
+	case *zng.TypeSet:
+		if !container {
+			return ErrSyntax
+		}
+		if err := t.set(v, body); err != nil {
+			return err
+		}
+	case *zng.TypeUnion:
+		if !container {
+			return ErrSyntax
+		}
+		if err := t.union(v, body); err != nil {
+			return err
+		}
+	default:
+		if container {
+			return ErrSyntax
+		}
+		if inner == t.Type {
+			t.value(body)
+		}
+	}
+	return nil
+}
+
+func (t *TypeIndexer) set(typ *zng.TypeSet, body zcode.Bytes) error {
+	if body == nil {
+		return nil
+	}
+	inner := zng.AliasedType(zng.InnerType(typ))
+	if zng.IsContainerType(inner) {
+		return ErrSyntax
+	}
+	if t.Type != inner {
+		return nil
+	}
+	it := zcode.Iter(body)
+	for !it.Done() {
+		body, container, err := it.Next()
+		if err != nil {
+			return err
+		}
+		if container {
+			return ErrSyntax
+		}
+		t.value(body)
+	}
+	return nil
+}
+
+func (t *TypeIndexer) record(typ *zng.TypeRecord, body zcode.Bytes) error {
+	if body == nil {
+		return nil
+	}
+	it := zcode.Iter(body)
+	for _, col := range typ.Columns {
+		if it.Done() {
+			return ErrSyntax
+		}
+		body, container, err := it.Next()
+		if err != nil {
+			return err
+		}
+		switch colType := zng.AliasedType(col.Type).(type) {
+		case *zng.TypeRecord:
+			if !container {
+				return ErrSyntax
+			}
+			if err := t.record(colType, body); err != nil {
+				return err
+			}
+		case *zng.TypeArray:
+			if !container {
+				return ErrSyntax
+			}
+			if err := t.vector(colType, body); err != nil {
+				return err
+			}
+		case *zng.TypeSet:
+			if !container {
+				return ErrSyntax
+			}
+			if err := t.set(colType, body); err != nil {
+				return err
+			}
+		case *zng.TypeUnion:
+			if !container {
+				return ErrSyntax
+			}
+			if err := t.union(colType, body); err != nil {
+				return err
+			}
+		default:
+			if container {
+				return ErrSyntax
+			}
+			if colType == t.Type {
+				t.value(body)
+			}
+		}
+	}
+	return nil
+}
+
+func (t *TypeIndexer) value(body zcode.Bytes) {
+	t.Table.Enter(string(body), nil)
```

Comment:
So the index simply says whether or not a value exists to its attached file. Isn't this the same as a perfectly accurate bloom filter that trades accuracy of results for space? Why not just use a bloom filter? What's the benefit?

It feels like if you are going the approach of creating indexes then it would make sense to produce a single index for a multiple files and the value of the index points to "hits". Otherwise you incur the cost of space for redundant key storage- something I image you would want to reduce.

Reply:
For low cardinality value sets, the overhead of redundant storage is far smaller than the size of the source data and the robustness and simplicity of having a separate index per archive file is very powerful.  For high cardinality value sets, the amount of overlap between files is usually going to be small (think of uids or file hashes) so the savings you would get from aggregation is small.  If it's not small, in worst case you are doubling the size of the data, but I think that will be rare in practice.  Bloom filters can definitely be useful here and we can get to that soon.

Reply:
Another consideration here regarding bloom filters is the ux for archive search involves returning the set of matched files so bloom-filter false positives could be counterintuitive.

Reply:
One advantage of separate indices is time-based retention: if you want to delete old data on a periodic basis, it's easy to drop "old data" by removing the index and zng for that old range, and not touch the rest of the data set.